```
tests/models/test_cleartasks.py::TestClearTasks::test_clear_task_instances: assert 3 == 1
 +  where 3 = <TaskInstance: test_clear_task_instances.1 scheduled__2016-01-01T00:00:00+00:00 [None]>.max_tries
tests/models/test_cleartasks.py::TestClearTasks::test_clear_task_instances_with_task_reschedule: AssertionError: assert 0 == 1
 +  where 0 = <function TestClearTasks.test_clear_task_instances_with_task_reschedule.<locals>.count_task_reschedule at 0x7f3f4c267b70>('1')
 +    where '1' = <TaskInstance: test_clear_task_instances_with_task_reschedule.1 scheduled__2016-01-01T00:00:00+00:00 [up_for_reschedule]>.task_id
tests/models/test_cleartasks.py::TestClearTasks::test_dag_clear: assert 3 == 1
 +  where 3 = <TaskInstance: test_dag_clear.test_dag_clear_task_1 scheduled__2016-01-01T00:00:00+00:00 [None]>.max_tries
```

Only failing on Postgres, so probably not related...? But DAG scheduling is so core I can't really be sure.

--- a/tests/jobs/test_scheduler_job.py
+++ b/tests/jobs/test_scheduler_job.py
@@ -3581,3 +3581,51 @@ class TestSchedulerJobQueriesCount:
                 assert start_date is None
                 assert end_date is None
                 assert duration is None
+
+    def test_catchup_works_correctly(self, dag_maker):
+        """Test that catchup works correctly"""
+        session = settings.Session()
+        with dag_maker(
+            dag_id='test_catchup_schedule_dag',
+            schedule_interval=timedelta(days=1),
+            start_date=DEFAULT_DATE,
+            catchup=True,
+            max_active_runs=1,
+            session=session,
+        ) as dag:
+            DummyOperator(task_id='dummy')
+
+        self.scheduler_job = SchedulerJob(subdir=os.devnull)
+        self.scheduler_job.executor = MockExecutor()
+        self.scheduler_job.processor_agent = mock.MagicMock(spec=DagFileProcessorAgent)
+
+        self.scheduler_job._create_dag_runs([dag_maker.dag_model], session)
+        self.scheduler_job._start_queued_dagruns(session)
+        # first dagrun execution date is DEFAULT_DATE 2016-01-01T00:00:00+00:00
+        dr = DagRun.find(execution_date=DEFAULT_DATE, session=session)[0]
+        ti = dr.get_task_instance(task_id='dummy')
+        ti.state = State.SUCCESS
+        session.merge(ti)
+        session.flush()
+
+        self.scheduler_job._schedule_dag_run(dr, session)
+        session.flush()
+
+        # Run the second time so _update_dag_next_dagrun will run
+        self.scheduler_job._schedule_dag_run(dr, session)
+        session.flush()
+
+        dag.catchup = False
+        dag.sync_to_db()
+        assert not dag.catchup
+
+        dm = DagModel.get_dagmodel(dag.dag_id)
+        self.scheduler_job._create_dag_runs([dm], session)
+
+        # Check catchup worked correctly by ensuring execution_date is quite new
+        # Our dag is a daily dag
+        assert (
+            session.query(DagRun.execution_date)
+            .filter(DagRun.execution_date != DEFAULT_DATE)  # exclude the first run
+            .scalar()
+        ) > (timezone.utcnow() - timedelta(days=2))

